### PR TITLE
clp 1.17.3 (new formula)

### PR DESCRIPTION
**Pull request by jamiesnape**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

File: Formula/clp.rb [hunk 1]
```diff
@@ -0,0 +1,125 @@
+class Clp < Formula
+  desc "Linear programming solver"
+  homepage "https://projects.coin-or.org/Clp"
+  url "https://www.coin-or.org/download/pkgsource/Clp/Clp-1.17.2.tgz"
+  sha256 "12320e817d9fdbaeef262bd76336002f65418f80ec295f21128acf0e833b283e"
+
+  depends_on "pkg-config" => :build
+  depends_on "gcc"
```
Comment (SMillerDev) [inline Formula/clp.rb]: during runtime too?

Comment (jamiesnape) [reply Formula/clp.rb]: Yes, for libgfortran.

File: Formula/clp.rb [hunk 2]
```diff
@@ -0,0 +1,125 @@
+class Clp < Formula
+  desc "Linear programming solver"
+  homepage "https://projects.coin-or.org/Clp"
+  url "https://www.coin-or.org/download/pkgsource/Clp/Clp-1.17.2.tgz"
```
Comment (soonho-tri) [inline Formula/clp.rb]: It seems that this link is not available. Can we use the one in github,  https://github.com/coin-or/Clp/archive/releases/1.17.2.zip ?

Comment (jamiesnape) [reply Formula/clp.rb]: Their whole website is down, actually. I'll see if it comes back up this morning.

Comment (soonho-tri) [reply Formula/clp.rb]: FYI, here is a version which uses GitHub links https://github.com/dreal-deps/homebrew-ibex/blob/master/clp%401.17.rb

**Comment by jamiesnape** [comment]
> * GitHub repository not notable enough (<30 forks, <30 watchers and <75 stars)

Note that this is a long-established project that recently moved (its mirror) to GitHub.

**Comment by soonho-tri** [comment]
> GitHub repository not notable enough (<30 forks, <30 watchers and <75 stars)

Also, this package is available in [Debian](https://packages.debian.org/search?suite=default&section=all&arch=any&searchon=names&keywords=coinor), [Ubuntu](https://launchpad.net/ubuntu/+source/clp), [Redhat](https://rpmfind.net/linux/rpm2html/search.php?query=coin-or-Clp), etc.

File: Formula/clp.rb [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        --disable-dependency-tracking
+        --disable-silent-rules
+        --prefix=#{prefix}
+      ]
+      system "./configure", *args
+      system "make"
+      system "make", "install"
+
+      # Remove redundant pkg-config file.
+      rm_f lib/"pkgconfig/coindatanetlib.pc"
```
Comment (fxcoudert) [inline Formula/clp.rb]: Why is it redundant? Is the issue reported upstream?

File: Formula/clp.rb [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        "--with-netlib-datadir=#{pkgshare}/coin/Data/Netlib",
+        "--with-blas-incdir=#{Formula["openblas"].opt_include}",
+        "--with-blas-lib=-L#{Formula["openblas"].opt_lib} -lopenblas",
+        "--with-lapack-incdir=#{Formula["openblas"].opt_include}",
+        "--with-lapack-lib=-L#{Formula["openblas"].opt_lib} -lopenblas",
+      ]
+      system "./configure", *args
+      system "make"
+
+      # Deparallelize due to error 1: "mkdir: #{include}/clp/coin: File exists."
```
Comment (fxcoudert) [inline Formula/clp.rb]: Has this been reported?

File: Formula/clp.rb [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+    args = [
+      "--datadir=#{pkgshare}",
+      "--disable-debug",
+      "--disable-dependency-tracking",
+      "--includedir=#{include}/clp",
+      "--prefix=#{prefix}",
+      "--with-blas-incdir=#{Formula["openblas"].opt_include}",
+      "--with-blas-lib=-L#{Formula["openblas"].opt_lib} -lopenblas",
+      "--with-coinutils-incdir=#{include}/clp/coin",
```
Comment (fxcoudert) [inline Formula/clp.rb]: Is there a simpler way to configure all this software together? If it's made to work altogether, it's surprising it needs so many different options to specify the subdirectories.

**Review by fxcoudert — CHANGES_REQUESTED** [review]

File: Formula/clp.rb (hunk 5 above)
Comment (jamiesnape) [reply Formula/clp.rb]: They are normally expecting a system-level install of all the dependencies. I am trying to keep them as self-contained as possible.

File: Formula/clp.rb (hunk 4 above)
Comment (jamiesnape) [reply Formula/clp.rb]: https://github.com/coin-or/Clp/issues/109

**Comment by jamiesnape** [comment]
Build failure has been resolved.

**Comment by tkralphs** [comment]
I just became aware of this submission because of [this PR](https://github.com/coin-or/Clp/issues/109). As documented in the Clp README, @dpo and I maintain a separate tap for COIN-OR projects, including Clp here:

https://github.com/coin-or-tools/homebrew-coinor

Clp has a number of dependencies, which should have their own formulae. This is because the [COIN-OR Optimization Suite](https://github.com/coin-or/COIN-OR-OptimizationSuite) contains multiple solvers with overlapping dependencies. In the Clp formula in this tap, we have put a lot of effort into supporting various build options, such as linking in [ampl-mp](https://github.com/Homebrew/homebrew-core/blob/master/Formula/ampl-mp.rb). All in all, it seems better not to add a separate duplicate formula here. We would be happy to collaborate on maintenance going forward.

**Comment by jamiesnape** [comment]
@tkralphs You don't provide bottles, so the tap is not useful for us.

**Comment by jamiesnape** [comment]
> Clp has a number of dependencies, which should have their own formulae.

I think that is a decision for the Homebrew maintainers. It could work either way.

> we have put a lot of effort into supporting various build options

Many users including ourselves do not need options. Homebrew itself has moved away from options. Bottles are much more important for our users.

> All in all, it seems better not to add a separate duplicate formula here. We would be happy to collaborate on maintenance going forward.

I strongly disagree. Making use of the community, experience, and infrastructure of core is a huge advantage.

**Comment by tkralphs** [comment]
@jamiesnape I am certainly not that knowledgeable about homebrew itself and all the issues involved, you are the expert there. But I do have intimate knowledge of COIN-OR solvers, our build system, and how it is evolving, as well as what default set of build options will probably cover the needs of the largest swath of users. The way you are building in this formula is deprecated and will stop working in the near future. 

If you would like to discuss the best way forward, capitalizing on our combined knowledge base, I would be happy to engage. If you prefer just to forge ahead, go for it. I just thought I would offer to help.

**Comment by jamiesnape** [comment]
What is deprecated about it?

**Comment by tkralphs** [comment]
In a nutshell, the `configure` script in the root directory is going away for a long list of reasons. It's a long story, but the individual dependencies will need to be built and installed one by one, not as a monolithic whole. There is a [new script](https://github.com/coin-or/coinbrew) for doing this and for managing all the dependencies when building from source. We can have a longer discussion about it off-line if you're interested, but this is part of the reason we have separate formulae for the dependencies in the tap. It works well, since homebrew already does all the dependency management.

We don't actually have much to do with maintaining the Debian packages, but we are in touch with the maintainers and they have already been doing it this way (separate package for each dependent project) for years.

**Comment by jamiesnape** [comment]
I'm happy for there to be separate packages for the dependencies if the Homebrew maintainers think that is acceptable.

Is the main objection here is the one package or that I am not using `coinbrew` (which my understanding is that Debian are not either)?

**Comment by tkralphs** [comment]
I don't think there's any way to use `coinbrew` in tandem with homebrew, so that is no problem. `coinbrew` is essentially doing the same job. If something like homebrew were supported on more platforms, we probably wouldn't need `coinbrew`. 

The biggest issue is that if you want to now add a formula for another COIN-OR solver, like [Cbc](https://github.com/coin-or/Cbc), that depends on some of the same packages (like [CoinUtils](https://github.com/coin-or/CoinUtils)), you end up with the same library being installed multiple times and there could even be version conflicts. In the past, this exact situation arose and is why we started down this road. 

As I said, I'm not actually much of an expert on homebrew, @dpo knows more and worked on the original monolithic formula that were subsequently removed.

**Comment by jamiesnape** [comment]
This would (roughly) be the split formula version: https://github.com/Homebrew/homebrew-core/compare/master...jamiesnape:clp-1.17.3-coinutils-2.11.2

**Comment by tkralphs** [comment]
If you don't want options, I would include `suite-sparse` for performance and `ampl-mp` (without it, you can't use Clp with some of the existing modeling system) by default. I personally would also make a formula for and include Osi, since that provides a useful C/C++ interface to the Clp library (and that of other solvers), plus a suite of unit tests. The unit tests require the full set of data instances so you would need a formula for installing that as well. But what you have seems to be sufficient for installing just the stand-alone Clp solver if you're not overly concerned about performance.

**Comment by jamiesnape** [comment]
I submitted #42739 toward having the dependencies as separate formulae.

**Comment by jamiesnape** [comment]
Since #42739 was merged, updated to use `coinutils` formula.

**Review by fxcoudert — APPROVED** [review]

**Pull request closed without merging**